lib/cocoapods/downloader/cache.rb

```diff
@@ -69,6 +69,93 @@ def cache_descriptors_per_pod
         end
       end
 
+      # Convenience method for acquiring a shared lock to safely read from the
+      # cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.read_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_SH, block)
+      end
+
+      # Convenience method for acquiring an exclusive lock to safely write to
+      # the cache. See `Cache.lock` for more details.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.write_lock(location, block = Proc.new)
+        Cache.lock(location, File::LOCK_EX, block)
+      end
+
+      # Creates a .lock file at `location`, aquires a lock of type
+      # `lock_type`, checks that it is valid, and executes `block` while holding
+      # on to that lock. Afterwards, the .lock file is deleted, which is why
+      # validation of the lock is necessary, as you might have a lock on a file
+      # that doesn't exist on the filesystem anymore.
+      #
+      # @param  [Pathname] location
+      #         the path to require a lock for.
+      #
+      # @param  [locking_constant] lock_type
+      #         the type of lock, either exclusive (File::LOCK_EX) or shared
+      #         (File::LOCK_SH).
+      #
+      # @param  [Proc] block
+      #         the block to execute inside the lock.
+      #
+      # @return [void]
+      #
+      def self.lock(destination, lock_type, block = Proc.new)
+        lockfile = "#{destination}.lock"
+        f = nil
+        loop do
+          f.close if f
+          f = File.open(lockfile, File::CREAT, 0o644)
+          f.flock(lock_type)
+          break if Cache.validate_lock(f, lockfile)
+        end
+        begin
+          block.call destination
```

Comment:
Yielding is generally considered to be the more performant option (or at least it was when I was still doing Ruby day to day).

```suggestion
          yield destination
```

Reply:
I was originally using yield, but then I wasn't able to pass the block between methods. Using `block = Proc.new` was the suggested solution that I found and it worked.

Reply:
When capturing the block (with the ampersand) it should work:

```
irb(main):002:0> def foo(&block)
irb(main):003:1>   bar(&block)
irb(main):004:1> end
=> :foo
irb(main):005:0> def bar(&block)
irb(main):006:1>   yield :ok
irb(main):007:1> end
=> :bar
irb(main):008:0> foo { |x| p x }
:ok
=> :ok
```

Reply:
I suppose I should try that out. :)

I was reading: https://joeyates.info/2014/06/30/how-to-forward-blocks-in-ruby/

In attempt 2, he was using the ampersand syntax, but said that "we can no longer pass a block to the generalized method", because it resulted in the following error:

    wrong number of arguments (1 for 2)

But he didn't use ampersand in both methods, so maybe that was the problem.

Actually just tried this and your way works fine. :)